pkg/pipeline/source/sdk/burst_estimator.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2025 LiveKit, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdk
+
+import (
+	"time"
+
+	"github.com/pion/rtp"
+	"github.com/pion/webrtc/v4"
+)
+
+type burstEstimator struct {
+	clockRate   uint32        // RTP clock rate for timestamp to duration conversion
+	maxSkew     time.Duration // tolerated skew between RTP duration and arrival delta
+	minArrival  time.Duration // lower bound on arrival delta to flag bursts
+	scoreTarget int           // number of consecutive stable packets needed
+	score       int           // running count of stable packets
+	lastTS      uint32        // previous RTP timestamp
+	lastArrival time.Time     // arrival time for the previous packet
+	hasLast     bool          // indicates lastTS/lastArrival already recorded
+	done        bool          // initialization deemed safe
+}
+
+func newBurstEstimator(clockRate uint32, kind webrtc.RTPCodecType) *burstEstimator {
+	be := &burstEstimator{
+		clockRate:   clockRate,
+		maxSkew:     20 * time.Millisecond,
+		minArrival:  2 * time.Millisecond,
+		scoreTarget: 5,
+	}
+
+	if kind == webrtc.RTPCodecTypeAudio {
+		be.maxSkew = 8 * time.Millisecond
+	}
+
+	return be
+}
+
+func (b *burstEstimator) ShouldInitialize(pkt *rtp.Packet, receivedAt time.Time) bool {
+	if b.done {
+		return true
+	}
+
+	if !b.hasLast {
+		b.lastTS = pkt.Timestamp
+		b.lastArrival = receivedAt
+		b.hasLast = true
+		return false
+	}
+
+	tsDelta := pkt.Timestamp - b.lastTS
+	arrivalDelta := receivedAt.Sub(b.lastArrival)
+
+	b.lastTS = pkt.Timestamp
+	b.lastArrival = receivedAt
```

Comment:
Should update this only on RTP time stamp change. As a frame could be broken up into several RTP packets which will all have same RTP timestamp, updating on every packet will skew this a bit towards detecting a burst for a large frame.

Reply:
And also should ignore negative `tsDelta`, i. e. out of order frames.